interfaces/apparmor/spec_test.go

```diff
@@ -144,26 +144,141 @@ apps:
   vanguard:
     command: vanguard
 layout:
-  /usr:
-    bind: $SNAP/usr
-  /mytmp:
+  /usr/foo:
+    bind: $SNAP/usr/foo
+  /var/tmp:
     type: tmpfs
     mode: 1777
-  /mylink:
-    symlink: $SNAP/link/target
+  /var/cache/mylink:
+    symlink: $SNAP_DATA/link/target
   /etc/foo.conf:
     bind-file: $SNAP/foo.conf
 `
 
 func (s *specSuite) TestApparmorSnippetsFromLayout(c *C) {
 	snapInfo := snaptest.MockInfo(c, snapWithLayout, &snap.SideInfo{Revision: snap.R(42)})
+	restore := apparmor.SetSpecScope(s.spec, []string{"snap.vanguard.vanguard"}, "vanguard")
+	defer restore()
+
 	s.spec.AddSnapLayout(snapInfo)
 	c.Assert(s.spec.Snippets(), DeepEquals, map[string][]string{
 		"snap.vanguard.vanguard": {
 			"# Layout path: /etc/foo.conf\n/etc/foo.conf mrwklix,",
-			"# Layout path: /mylink\n# (no extra permissions required for symlink)",
-			"# Layout path: /mytmp\n/mytmp{,/**} mrwklix,",
-			"# Layout path: /usr\n/usr{,/**} mrwklix,",
+			"# Layout path: /usr/foo\n/usr/foo{,/**} mrwklix,",
+			"# Layout path: /var/cache/mylink\n# (no extra permissions required for symlink)",
+			"# Layout path: /var/tmp\n/var/tmp{,/**} mrwklix,",
 		},
 	})
+
+	profile0 := `  # Layout /etc/foo.conf: bind-file $SNAP/foo.conf
+  mount options=(bind, rw) /snap/vanguard/42/foo.conf -> /etc/foo.conf,
+  umount /etc/foo.conf,
+  # Writable mimic /etc
+  mount options=(rbind, rw) /etc/ -> /tmp/.snap/etc/,
+  mount fstype=tmpfs options=(rw) tmpfs -> /etc/,
+  mount options=(rbind, rw) /tmp/.snap/etc/** -> /etc/**,
+  mount options=(bind, rw) /tmp/.snap/etc/* -> /etc/*,
+  umount /tmp/.snap/etc/,
+  umount /etc{,/**},
+  /etc/** rw,
+  /etc/ rw,
+  /tmp/.snap/etc/** rw,
+  /tmp/.snap/etc/ rw,
+  /tmp/.snap/ rw,
+  /tmp/ rw,
+  # Writable mimic /snap/vanguard/42
+  mount options=(rbind, rw) /snap/vanguard/42/ -> /tmp/.snap/snap/vanguard/42/,
+  mount fstype=tmpfs options=(rw) tmpfs -> /snap/vanguard/42/,
+  mount options=(rbind, rw) /tmp/.snap/snap/vanguard/42/** -> /snap/vanguard/42/**,
+  mount options=(bind, rw) /tmp/.snap/snap/vanguard/42/* -> /snap/vanguard/42/*,
+  umount /tmp/.snap/snap/vanguard/42/,
+  umount /snap/vanguard/42{,/**},
+  /snap/vanguard/42/** rw,
+  /snap/vanguard/42/ rw,
+  /snap/vanguard/ rw,
+  /snap/ rw,
+  /tmp/.snap/snap/vanguard/42/** rw,
+  /tmp/.snap/snap/vanguard/42/ rw,
+  /tmp/.snap/snap/vanguard/ rw,
+  /tmp/.snap/snap/ rw,
+  /tmp/.snap/ rw,
+  /tmp/ rw,
```

Comment:
Huh. I didn't expect snap-update-ns to need to be able to mkdir(/tmp)...

Reply:
Technically it will try if /tmp is empty. I think we can special case this and have a hard-coded `/tmp/ r` rule.

Reply:
What do you mean be 'if /tmp is empty'? /tmp doesn't exist or /tmp exists but is empty? The rw rule would be only for if /tmp doesn't exist, which is what '/tmp/ rw,' allows. If /tmp is empty, then you have rules like '/tmp/foo rw,'.

Reply:
Er, I meant if /tmp doesn't exist.